[Viewport Clipping] [Part 4/4] Add the ability to clip fixed and sticky elements to the viewport

whsieh opened the pull request:
#### 27e72ea1a69bee0f87ea794f1f55e34c7d7c4650
<pre>
[Viewport Clipping] [Part 4/4] Add the ability to clip fixed and sticky elements to the viewport
<a href="https://bugs.webkit.org/show_bug.cgi?id=292125">https://bugs.webkit.org/show_bug.cgi?id=292125</a>
<a href="https://rdar.apple.com/147964717">rdar://147964717</a>

Reviewed by Abrar Rahman Protyasha and Simon Fraser.

Finish support for clipping fixed and sticky elements to the viewport, by (finally) introducing the
viewport clipping layer to `RenderLayerBacking`. This clipping layer sits above the current viewport
anchor layer (i.e. the transform layer) in the compositing hierarchy, and clips all fixed or sticky
content to the bounds of the viewport. This clipping layer is installed:

1. Only for fixed elements in the main frame, or
2. Sticky elements in the main frame that aren&apos;t also contained within an overflow scroller.

...and additionally requires the `ContentInsetBackgroundFill` setting to be enabled.

When scrolling with a viewport clipping layer, the fixed or sticky nodes in the scrolling tree now
blit the clipping layer instead of the anchor layer; in the case of sticky nodes, this also means we
counter-blit the anchor layer underneath as needed (in the case where the sticky node still scrolls
with page content instead of behaving as fixed).

See below for more details.

* LayoutTests/TestExpectations:
* LayoutTests/compositing/cocoa/clip-fixed-element-to-viewport-expected.html: Added.
* LayoutTests/compositing/cocoa/clip-fixed-element-to-viewport.html: Added.
* LayoutTests/compositing/cocoa/clip-sticky-element-to-viewport-expected.html: Added.
* LayoutTests/compositing/cocoa/clip-sticky-element-to-viewport.html: Added.

Add new ref tests to verify that fixed and sticky elements are clipped to the top of the viewport
when scrolling. The test pages both contain fixed and sticky elements that partially lie outside of
the viewport, while the refs contain fixed and sticky elements that are fully contained in the
viewport.

* LayoutTests/compositing/ios/immediate-scrolling-clip-fixed-element-to-viewport-expected.html: Added.
* LayoutTests/compositing/ios/immediate-scrolling-clip-fixed-element-to-viewport.html: Added.
* LayoutTests/compositing/ios/immediate-scrolling-clip-sticky-element-to-viewport-expected.html: Added.
* LayoutTests/compositing/ios/immediate-scrolling-clip-sticky-element-to-viewport.html: Added.

Add more ref tests to exercise the new clipping behavior after immediate scrolling, when scrolling
updates are disabled.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateDebugIndicators):

Add a debug indicator to the new `m_viewportClippingLayer`.

(WebCore::RenderLayerBacking::destroyGraphicsLayers):
(WebCore::RenderLayerBacking::updateConfiguration):
(WebCore::RenderLayerBacking::updateGeometry):

Account for the presence of the viewport clipping layer by setting its size and position based on
the layout viewport rect.

(WebCore::RenderLayerBacking::updateInternalHierarchy):

Add the clipping layer as an ordered layer, above the existing viewport anchor layer.

(WebCore::RenderLayerBacking::updateViewportConstrainedSublayers):

Rename this to `-Sublayers`, to reflect the fact that this now updates both the viewport clipping
layer and the viewport anchor layer. Instead of a bool, this now also takes a new enum that
represents which layers are needed (i.e. anchor, clipping + anchor, or neither). See below for more
details.

(WebCore::RenderLayerBacking::updateViewportConstrainedAnchorLayer): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.h:

Add `m_viewportClippingLayer`, and make `viewportClippingOrAnchorLayer()` return either the clipping
layer (if it exists) or the anchor layer.

* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::viewportConstrainedSublayers const):

Replaces `isViewportConstrainedFixedOrStickyLayer`; see below for more details.

(WebCore::RenderLayerCompositor::coordinatedScrollingRolesForLayer const):
(WebCore::RenderLayerCompositor::isViewportConstrainedFixedOrStickyLayer const): Deleted.

Rename this to `viewportConstrainedSublayers`. Instead of returning a `bool`, return a new enum that
represents whether the `RenderLayer` should contain (above its child content layer):

- None                 no viewport constrained layers (e.g. not fixed or sticky).
- Anchor               a single viewport anchor layer (transform layer)
- ClippingAndAnchor    a viewport-sized clipping layer, with the clipping layer underneath.

* Source/WebCore/rendering/RenderLayerCompositor.h:
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::TestController::platformResetStateToConsistentValues):

Drive-by fix: check all content inset edges (not just the top) when resetting state between tests.

* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::setObscuredInsets):

Adjust this testing helper to additionally set content insets and adjust layout viewport dimensions
to better match obscured inset behavior in Safari and other apps.

Canonical link: <a href="https://commits.webkit.org/294368@main">https://commits.webkit.org/294368@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/108e60c845fece2a157c3422bea162aa6ee586ab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101575 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21243 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11554 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106733 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52208 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103615 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21551 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29738 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77355 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34384 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104582 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16642 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91730 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57692 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16467 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9747 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51555 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86335 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9824 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109084 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28708 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/21128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86331 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29069 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87931 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85893 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21862 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30638 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8348 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22842 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28636 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33925 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28447 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31770 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30006 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -2596,9 +2611,28 @@ bool RenderLayerBacking::updateTransformFlatteningLayer(const RenderLayer* compo
+        needsClippingLayer = true;
+        break;
+    }
+    }
```
Maybe some fallthrough so that DRY? Feel free to ignore if you think it's confusing.

Also, you don't need new scopes if you aren't declaring variables.

```suggestion
    switch (sublayers) {
    case NoLayers:
        break;
    case ClippingAndAnchorLayers:
        needClippingLayer = true;
        FALLTHROUGH;
    case AnchorLayer:
        needsAnchorLayer = true;
        break;
    }
```

whsieh replied:
Sounds good to me — will update this switch case.

Thanks for the review!

aprotyas commented on Source/WebCore/rendering/RenderLayerCompositor.h:
```diff
@@ -100,6 +100,12 @@ enum class ScrollCoordinationRole {
+    NoLayers,
+    AnchorLayer,
+    ClippingAndAnchorLayers,
+};
```
Optional: does the `Layer[s]` prefix in the case names help with readability? If not, maybe just:

```suggestion
enum class ViewportConstrainedSublayers : uint8_t {
    None,
    Anchor,
    ClippingAndAnchor,
};
```

whsieh replied:
Makes sense!

aprotyas reviewed (approved):
LGTM, mostly cosmetic comments

smfr commented on LayoutTests/compositing/ios/immediate-scrolling-clip-sticky-element-to-viewport-expected.html:
```diff
@@ -0,0 +1,49 @@
+    await UIHelper.setObscuredInsets(50, 0, 0, 0);
+    await UIHelper.ensurePresentationUpdate();
+    await UIHelper.immediateScrollElementAtContentPointToOffset(100, 100, 0, 400, true);
+    await UIHelper.ensurePresentationUpdate();
```
Does this negate the "don't send updates from the UI to the web process"? I'm not sure.

whsieh replied:
Upon further testing, it looks like this presentation update isn't necessary here at all. Removed!

(FWIW, it didn't seem like ensuring a presentation update here changed anything w.r.t. scrolling state node updates, based on logging in `reconcileLayerPositionForViewportRect`)

smfr commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -531,6 +531,11 @@ void RenderLayerBacking::updateDebugIndicators(bool showBorder, bool showRepaint
 
+    if (m_viewportClippingLayer) {
+        m_viewportClippingLayer->setShowDebugBorder(showBorder);
+        m_viewportClippingLayer->setShowRepaintCounter(showRepaintCounter);
```
It will never show a repaint counter.

whsieh replied:
Oh, right — removed!

whsieh committed:
[Viewport Clipping] [Part 4/4] Add the ability to clip fixed and sticky elements to the viewport
https://bugs.webkit.org/show_bug.cgi?id=292125
rdar://147964717

Reviewed by Abrar Rahman Protyasha and Simon Fraser.

Finish support for clipping fixed and sticky elements to the viewport, by (finally) introducing the
viewport clipping layer to `RenderLayerBacking`. This clipping layer sits above the current viewport
anchor layer (i.e. the transform layer) in the compositing hierarchy, and clips all fixed or sticky
content to the bounds of the viewport. This clipping layer is installed:

1. Only for fixed elements in the main frame, or
2. Sticky elements in the main frame that aren't also contained within an overflow scroller.

...and additionally requires the `ContentInsetBackgroundFill` setting to be enabled.

When scrolling with a viewport clipping layer, the fixed or sticky nodes in the scrolling tree now
blit the clipping layer instead of the anchor layer; in the case of sticky nodes, this also means we
counter-blit the anchor layer underneath as needed (in the case where the sticky node still scrolls
with page content instead of behaving as fixed).

See below for more details.

* LayoutTests/TestExpectations:
* LayoutTests/compositing/cocoa/clip-fixed-element-to-viewport-expected.html: Added.
* LayoutTests/compositing/cocoa/clip-fixed-element-to-viewport.html: Added.
* LayoutTests/compositing/cocoa/clip-sticky-element-to-viewport-expected.html: Added.
* LayoutTests/compositing/cocoa/clip-sticky-element-to-viewport.html: Added.

Add new ref tests to verify that fixed and sticky elements are clipped to the top of the viewport
when scrolling. The test pages both contain fixed and sticky elements that partially lie outside of
the viewport, while the refs contain fixed and sticky elements that are fully contained in the
viewport.

* LayoutTests/compositing/ios/immediate-scrolling-clip-fixed-element-to-viewport-expected.html: Added.
* LayoutTests/compositing/ios/immediate-scrolling-clip-fixed-element-to-viewport.html: Added.
* LayoutTests/compositing/ios/immediate-scrolling-clip-sticky-element-to-viewport-expected.html: Added.
* LayoutTests/compositing/ios/immediate-scrolling-clip-sticky-element-to-viewport.html: Added.

Add more ref tests to exercise the new clipping behavior after immediate scrolling, when scrolling
updates are disabled.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateDebugIndicators):

Add a debug indicator to the new `m_viewportClippingLayer`.

(WebCore::RenderLayerBacking::destroyGraphicsLayers):
(WebCore::RenderLayerBacking::updateConfiguration):
(WebCore::RenderLayerBacking::updateGeometry):

Account for the presence of the viewport clipping layer by setting its size and position based on
the layout viewport rect.

(WebCore::RenderLayerBacking::updateInternalHierarchy):

Add the clipping layer as an ordered layer, above the existing viewport anchor layer.

(WebCore::RenderLayerBacking::updateViewportConstrainedSublayers):

Rename this to `-Sublayers`, to reflect the fact that this now updates both the viewport clipping
layer and the viewport anchor layer. Instead of a bool, this now also takes a new enum that
represents which layers are needed (i.e. anchor, clipping + anchor, or neither). See below for more
details.

(WebCore::RenderLayerBacking::updateViewportConstrainedAnchorLayer): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.h:

Add `m_viewportClippingLayer`, and make `viewportClippingOrAnchorLayer()` return either the clipping
layer (if it exists) or the anchor layer.

* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::viewportConstrainedSublayers const):

Replaces `isViewportConstrainedFixedOrStickyLayer`; see below for more details.

(WebCore::RenderLayerCompositor::coordinatedScrollingRolesForLayer const):
(WebCore::RenderLayerCompositor::isViewportConstrainedFixedOrStickyLayer const): Deleted.

Rename this to `viewportConstrainedSublayers`. Instead of returning a `bool`, return a new enum that
represents whether the `RenderLayer` should contain (above its child content layer):

- None                 no viewport constrained layers (e.g. not fixed or sticky).
- Anchor               a single viewport anchor layer (transform layer)
- ClippingAndAnchor    a viewport-sized clipping layer, with the clipping layer underneath.

* Source/WebCore/rendering/RenderLayerCompositor.h:
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:
(WTR::TestController::platformResetStateToConsistentValues):

Drive-by fix: check all content inset edges (not just the top) when resetting state between tests.

* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::setObscuredInsets):

Adjust this testing helper to additionally set content insets and adjust layout viewport dimensions
to better match obscured inset behavior in Safari and other apps.

Canonical link: https://commits.webkit.org/294368@main
